src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,130 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double ADD_CAPACITY = 1.5;
+    private static final int POSITION_ZERO = 0;
+    private T [] defaultArray;
+    private int listSize;
+
+    public ArrayList() {
+        defaultArray = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void isIndexInBounds(int index) {
+        if (!(index >= 0 && index < listSize)) {
+            throw new ArrayListIndexOutOfBoundsException("Index: "
+                    + index
+                    + " is out of bounds. "
+                    + "Size: "
+                    + listSize);
+        }
+    }
+
+    private void isIndexInRangeInclusive(int index) {
+        if (!(index >= 0 && index <= listSize)) {
+            throw new ArrayListIndexOutOfBoundsException("Index: "
+                    + index
+                    + " is out of bounds. "
+                    + "Size: "
+                    + listSize);
+        }
+    }
+
+    private void growList() {
+        if (listSize == defaultArray.length) {
+            Object [] extendedArray = new Object[(int) (defaultArray.length * ADD_CAPACITY)];
+            System.arraycopy(defaultArray,
+                    POSITION_ZERO,
+                    extendedArray,
+                    POSITION_ZERO,
+                    defaultArray.length);
+            defaultArray = (T[]) extendedArray;
+        }
```

Comment:
The 'growList' method can be improved by encapsulating the resizing logic in a separate method, as the current implementation mixes the check for size with the actual resizing. Consider creating a method like 'ensureCapacity' that only resizes the array if necessary.